src/app/components/chat/window/output/AppChatWindowOutputItem.vue

```diff
@@ -0,0 +1,530 @@
+const isEditingColor = computed(() => {
+	if (!isEditing.value) {
+		return undefined;
+	}
+
+	const highlight = '#' + actualTheme.value.highlight_;
+	const backlight = '#' + actualTheme.value.backlight_;
+
+	const tintColor = isDark.value ? highlight : backlight;
+	return transparentize(0.65, tintColor);
+});
+
+const editingState = computed(() => {
+	if (isEditing.value) {
+		return {
+			/** The text to display in the comment */
+			display: '(editing...)',
+			/** The tooltip to display on hover or tap */
+			tooltip: '',
+		};
+	}
+
+	if (message.value.edited_on) {
+		return {
+			display: '(edited)',
+			tooltip: formatDate(message.value.edited_on!, 'medium'),
+		};
+	}
+
+	return null;
+});
+
+const shouldShowMessageOptions = computed(() => canRemoveMessage.value || canEditMessage.value);
+
+const dataAnchorWidth = computed(() => {
+	const itemWidth = 40;
+	if (isEditing.value) {
+		return itemWidth;
+	}
+
+	let itemCount = 0;
+	if (canRemoveMessage.value) {
+		++itemCount;
+	}
+	if (canEditMessage.value) {
+		++itemCount;
+	}
+
+	return itemWidth * itemCount + (itemCount - 1);
+});
+
+const canRemoveMessage = computed(() => {
+	if (!chat.value.currentUser) {
+		return false;
+	}
+
+	// The owner of the message can remove it.
+	if (chat.value.currentUser.id === message.value.user.id) {
+		return true;
+	}
+
+	// Mods/Room owners can also remove the message.
+	return userCanModerateOtherUser(
+		chat.value,
+		room.value,
+		chat.value.currentUser,
+		message.value.user
+	);
+});
+
+const canEditMessage = computed(() => {
+	// Only content messages can be edited.
+	if (message.value.type !== 'content') {
+		return false;
+	}
+
+	if (!chat.value.currentUser) {
+		return false;
+	}
+
+	// Only the owner of the message can edit.
+	return chat.value.currentUser.id === message.value.user.id;
+});
+
+function startEdit() {
+	setMessageEditing(chat.value, message.value);
+	Popper.hideAll();
+}
+
+function stopEdit() {
+	setMessageEditing(chat.value, null);
+	Popper.hideAll();
+}
+
+async function removeMessage() {
+	Popper.hideAll();
+
+	const result = await ModalConfirm.show(
+		$gettext(`Are you sure you want to remove this message?`)
+	);
+
+	if (!result) {
+		return;
+	}
+
+	setMessageEditing(chat.value, null);
+	chatRemoveMessage(chat.value, room.value, message.value.id);
+}
+
+let canClearFocus = false;
+let isFocused = false;
+
+async function onFocusItem() {
+	isFocused = true;
+	// Allow a second click to blur only after ~1/2 of our animations play.
+	setTimeout(() => (canClearFocus = isFocused), 100);
+}
+
+function onBlurItem() {
+	isFocused = false;
+	canClearFocus = false;
+}
+
+async function onClickItem() {
+	if (isFocused && canClearFocus) {
+		itemWrapper.value?.blur();
+	}
+}
+
+function onItemWrapperResize() {
+	if (!itemWrapper.value) {
+		return;
+	}
+
+	const { offsetWidth } = itemWrapper.value;
+	const decoratorPadding = 12 * 2;
+	contentViewerBounds.width = Math.max(offsetWidth - decoratorPadding, 100);
```

Comment:
Will this be the same for every item? It seems like a lot do for each item added into the chat when the value is the same. Do you think we can just put a watcher on the AppChatWindowOutput component and then subtract the correct spacing and pass it into each chat item?

Reply:
@hapurubokka Have these items instead take props for both their maxWidth, and their left-padding used for the avatar.